src/aleph/web/controllers/storage.py

```diff
@@ -56,20 +92,133 @@ async def add_storage_json_controller(request: web.Request):
     return web.json_response(output)
 
 
+async def _verify_message_signature(
+    pending_message: BasePendingMessage, chain_service: ChainService
+) -> None:
+    try:
+        await chain_service.verify_signature(pending_message)
+    except InvalidSignature:
+        raise web.HTTPForbidden()
+
+
+async def _verify_user_balance(
+    pending_message_db: PendingMessageDb, session: DbSession, size: int
+) -> None:
+    current_balance = get_total_balance(
+        session=session, address=pending_message_db.sender
+    ) or Decimal(0)
+    required_balance = (size / MiB) / 3
+    current_cost_for_user = get_total_cost_for_address(
+        session=session, address=pending_message_db.sender
+    )
+    if current_balance < (Decimal(required_balance) + current_cost_for_user):
+        raise web.HTTPPaymentRequired
+
+
+async def _verify_user_file(message: PendingStoreMessage, size: int, file_io) -> None:
+    file_io.seek(0)
```

Comment:
Why do you need to seek? Are you reading the file twice?

Reply:
[            file_content = fileobject.read()](https://github.com/1yam/pyaleph/blob/4db385efd6ee8e1d2677091cb8c13a36e283272d/src/aleph/storage.py#L274)

The functions add_file on Storage Service class wasn't doing seek when after read (he was doing it for ipf,  I replace the seek here and not on _verify_user_file

Reply:
We really shouldn't read the file twice, that's an expensive operation.